cli/listener/src/main.rs

```diff
@@ -165,46 +170,55 @@ pub fn get_interval_price_for_period(
-              SELECT
-                *
-              FROM
-                prices_epoch
-              ORDER BY
-                timestamp ASC
-            ",
+        // Get minimum epoch in which timestamp is greater than `from_time`.
+        let mut first_epoch_stmt = tx.prepare(
+            "SELECT MIN(epoch) from exchange_rate where pool = :pool and timestamp > :t",
+        )?;
+        let epoch = match first_epoch_stmt
+            .query_row([pool.clone(), from_time.to_rfc3339()], |row| {
+                row.get::<usize, u64>(0)
+            })
+            .ok()
+        {
+            Some(epoch) => epoch,
+            None => return Ok(None),
+        };
+        let minimum_slot = epoch_schedule.get_first_slot_in_epoch(epoch) + QUERY_SLOT_OFFSET;
+
+        // Get the first row from `epoch` in which the slot is greater than `minimum_slot`.
+        let mut exchange_rate_stmt = tx.prepare(
+            "SELECT * from exchange_rate WHERE epoch = :epoch AND slot >= :slot_min LIMIT 1",
+        )?;
+        let first_exchange_rate = exchange_rate_stmt
+            .query_map([epoch, minimum_slot], row_map)?
+            .next();
+
+        // Get maximum epoch in which timestamp is smaller than `from_time`.
+        let mut last_epoch_stmt = tx.prepare(
+            "SELECT MAX(epoch) from exchange_rate where pool = :pool and timestamp < :t",
+        )?;
+        let epoch = match last_epoch_stmt
+            .query_row([pool, to_time.to_rfc3339()], |row| row.get::<usize, u64>(0))
+            .ok()
+        {
+            Some(epoch) => epoch,
+            None => return Ok(None),
+        };
+        let minimum_slot = epoch_schedule.get_first_slot_in_epoch(epoch) + QUERY_SLOT_OFFSET;
+
+        // Get the first row from `epoch` in which the slot is greater than `minimum_slot`.
+        let mut exchange_rate_stmt = tx.prepare(
+            "SELECT * from exchange_rate WHERE epoch = :epoch AND slot >= :slot_min LIMIT 1",
         )?;
-        // Get first logged maximal logged data based on timestamp that is smaller than `to_time`.
-        let stmt_last =
-            &mut tx.prepare("WITH prices_epoch AS (
-                SELECT *
-                FROM exchange_rate
-                WHERE epoch = (SELECT MAX(epoch) from exchange_rate where pool = :pool AND timestamp < :t)
-              )
-              SELECT
-                *
-              FROM
-                prices_epoch
-              ORDER BY
-                timestamp ASC
-            ")?;
-        let mut row_iter = stmt_first.query_map([pool.clone(), from_time.to_rfc3339()], row_map)?;
-        let first = row_iter.next();
-
-        let mut row_iter = stmt_last.query_map([pool, to_time.to_rfc3339()], row_map)?;
-        let last = row_iter.next();
-
-        (first, last)
+        let last_exchange_rate = exchange_rate_stmt
+            .query_map([epoch, minimum_slot], row_map)?
+            .next();
+
+        (first_exchange_rate, last_exchange_rate)
```

Comment:
To factiltate these queries, we should create an index either on `timestamp` and `slot`, or on `(pool, timestamp)` and `(pool, slot)`. Otherwise it will need to do a full table scan every time. Probably the overhead of the extra column in the index is not worth scanning through a few extra rows to find a matching `pool`, I would keep it simple and index just `timestamp` and `slot`.

We can add this to the startup:

```sql
CREATE INDEX IF NOT EXISTS ix_exchange_rate_timestamp ON exchange_rate (timestamp);
CREATE INDEX IF NOT EXISTS ix_exchange_rate_slot ON exchange_rate (slot);
```

Reply:
Yeah, great idea :)